File: app/src/lib/logger.ts
interface Logger {
  logError(err: unknown): void;
}

class ConsoleLogger implements Logger {
  constructor(private name: string) {}

  logError(err: unknown): void {
    console.error({
      name: this.name,
      error: err,
    });
  }
}

export function createDefaultLogger(name: string): Logger {
  return new ConsoleLogger(name);
}

File: api/lib/logger.go
package lib

import (
	"fmt"
)

type (
	Logger interface {
		LogError(err error)
	}

	LoggerWithConfig struct {
		Logger
		name string
	}

	NewLoggerFunc func(name string) Logger
)

var NewLogger NewLoggerFunc = func(name string) Logger {
	return LoggerWithConfig{
		name: name,
	}
}

func (logger LoggerWithConfig) LogError(err error) {
	fmt.Printf("From %s:\n%+v", logger.name, err)
}
